### PR TITLE
SQL Clients memory usage and performance improvements

### PR DESCRIPTION
**Pull request by tsegismont**
A series of improvements to memory usage of SQL clients in general and MySQL in particular.

Commits can be reviewed individually.

### CODE REVIEW

**Comment by tsegismont** [comment]
cc @franz1981 (I think you've reviewed my working branch already but here's the proposed change to Vert.x 4.3)

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowSetImpl.java
Comment (franz1981) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowSetImpl.java:81]: Three different types here is bad for the JIT: better 2 and `null` if empty or just 2  (choosing the most common ones).

File: vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorReadBenchmark.java
Comment (franz1981) [inline vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorReadBenchmark.java:30]: Will run today I promise :) I'm fighting with microbenchs as well recently

**Review by franz1981 — CHANGES_REQUESTED** [review]
Just minor changes, great work!
Feel free to past some of the results after applying this, just to make easier for the future ourselves remember how good the results are :)

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:54]: No, it is not. I guess `ArrayList` has it because the index could be within the range of capacity but not of the size.

I'll remove it.

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:64]: Not sure what you mean, it extends `AbstractList` which throws `UnsupportedOperationException` on modifiers.

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/CommandCodec.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/CommandCodec.java:158]: I'll reduce the number of invocations.

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java:28]: ok

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (franz1981) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:64]: If outside it's used as `ColumnNames` and never as `AbstractList` (if not for few/one single trivial method) you can save extending it.
It will make the JIT live a bit easier (not another list implementor bla bla bla) - but I'm ok whatever choice you made here, really

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowSetImpl.java
Comment (tsegismont) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowSetImpl.java:81]: ok

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:61]: Outside it's never used as `ColumnNames`, only as `List`:

https://github.com/eclipse-vertx/vertx-sql-client/blob/cf1da8a1652f5d1cefaec7c17cd892bfbb4d1448/vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java#L44-L46

https://github.com/eclipse-vertx/vertx-sql-client/blob/cf1da8a1652f5d1cefaec7c17cd892bfbb4d1448/vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java#L33-L34

https://github.com/eclipse-vertx/vertx-sql-client/blob/3b303d98f2410455eb47f01ac54d03f1c3dfccdf/vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowDesc.java#L62-L65

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java
Comment (vietj) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java:35]: it is not useful because that works mainly with servers, on a client you write data first and then read it.

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (vietj) [inline vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:62]: eventually this should be moved to `RowDesc` is a follow up PR

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (vietj) [inline vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:93]: eventually this should be moved to `RowDesc` is a follow up PR

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (vietj) [inline vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:31]: it does not seem worth to be public instead of using the accessor

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java
Comment (vietj) [inline vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java:23]: you should look at the PgDecoder that is very similar, it is written in a way that it will slice packet itself without using `LengthFieldBasedFrameDecoder` which saves an inbound handler in the channel pipeline.

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/protocol/ColumnDefinition.java
Comment (vietj) [inline vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/protocol/ColumnDefinition.java:24]: add a comment saying that this is a subset of the data that we actually need in practice

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:62]: In fact, I may do this when applying changes to other clients.

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java:93]: In fact, I may do this when applying changes to other clients.

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/Accumulator.java
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/Accumulator.java:21]: should be named `RowAccumulator`

**Review by vietj — CHANGES_REQUESTED** [review]
a few comments inline

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/CommandCodec.java
Comment (franz1981) [inline vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/CommandCodec.java:166]: lovely, you saved many bound checks here by having just 2 skipBytes

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java
Comment (franz1981) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java:32]:
```suggestion
  public Object get(PropertyKind<?> property) {
      final Object[] elements = this.elements;
      for (int i = 0, size = elements.length; i < size; i+=2) {
         final Object key = elements[i];
         assert key != null : "key cannot be null"; 
         if (key == property || key.equals(property)) {
             return elements[i + 1];
         }  
      }
      return null;
  }
```
This will help JIT to correlate the loop variables with the length of the array, saving bound checks and inlining the equality check would help in case key is effectively the same (if it's not common, you can drop the ref equality check)

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java
Comment (franz1981) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java:43]: I would perform the same loop as in the previous comment, that help reduce bound checks

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java
Comment (franz1981) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java:56]: count is that necessary?
you can just use elements.length / 2

File: vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorReadBenchmark.java
Comment (franz1981) [inline vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorReadBenchmark.java:59]: you can just consume `s`?

Comment (franz1981) [reply vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorReadBenchmark.java:59]: or you can do something better here: `BlackHole::consume` preven some JIT loop transformations that are key to compare between versions:

```java
  @Benchmark
  public int iterate() {
    int dummy = 0;
    for (String s : accumulator) {
      dummy += s.length();
    }
    return dummy;
}      
```
This will make the loop as fast as it can as in the real world and will prevent it to be dead-code-eliminated

I suggest to re-run the bench after changing this

File: vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorWriteBenchmark.java
Comment (franz1981) [inline vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorWriteBenchmark.java:29]: value >= 2

**Review by franz1981 — CHANGES_REQUESTED** [review]
LGTM minor changes

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java
Comment (tsegismont) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java:56]: right, I'll remove it

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java
Comment (tsegismont) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/PropertyKindMap.java:32]: I just pushed 8620798d34eb9f84942f2a13f78bfa33d3d4831d to remove the computed `count` field. At the same time, I've changed the loop to use the `elements.length` in condition check. Can you verify it's enough or is still better to make a local copy of the reference to the array?

File: vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorWriteBenchmark.java
Comment (tsegismont) [reply vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorWriteBenchmark.java]: Sorry, can you explain why it is needed?

Comment (tsegismont) [reply vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorWriteBenchmark.java]: I found this https://stackoverflow.com/a/35147232/2133695

In 97039b66 I've improved the benchmarks. I'll run them tonight because the ETA with 3 forks is more than 7 hours ahead :sweat_smile:

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/codec/MySQLDecoder.java:23]: Let's address that in a separate PR. I got rid of the last `slice()` calls in this decoder in 36574ee1.

File: vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorWriteBenchmark.java
Comment (tsegismont) [reply vertx-sql-client/src/test/java/io/vertx/sqlclient/benchmarks/AccumulatorWriteBenchmark.java]: I've rolled-back to `ArrayListRowAccumulator` in edb479f

Latest test results do not demonstrate good results for `ChunkedRowAccumulator`.
Until we sort this out, let's continue with `ArrayListRowAccumulator`.

The biggest improvement here anyway is to avoid creating an accumulator if the database returns no rows or a single (which happens very often).

**Comment by tsegismont** [comment]
@vietj @franz1981 PTAL

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/ChunkedRowAccumulator.java
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/ChunkedRowAccumulator.java:19]: do we actually use this in our code ?

Comment (tsegismont) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/ChunkedRowAccumulator.java:19]: Not yet. I've met with @franz1981 this morning to talk about the last benchmarks results I got. I will run another round of tests this afternoon.

So we can keep it until we make a final decision. But let's not block merging of this PR because of this.

Comment (vietj) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/ChunkedRowAccumulator.java:19]: sure, but I'm not convinced that the complexity it brings is worth it

Comment (tsegismont) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/ChunkedRowAccumulator.java:19]: We'll see what the new results say. It won't be a game changer regarding latency but allocations may be reduced significantly.

Any other comments about the recent commits?

Comment (vietj) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/accumulator/ChunkedRowAccumulator.java:19]: all seem fine, we'll rediscuss th chunked accumulaor later then

**Review by vietj — APPROVED** [review]

**Pull request merged**
